File: examples/showcase/slack_fronting/custom-code-bundles/slack_fronting_csharp/BackendAdapter.cs
using System.Text.Json;
using Anip.Service;

namespace {{ANIP_CSHARP_ROOT_NAMESPACE}};

public delegate Dictionary<string, object?> BackendAdapterHandler(
    Dictionary<string, object?> capability,
    Dictionary<string, object?> plan,
    Dictionary<string, object?> adapterInput,
    InvocationContext context);

public static class BackendAdapter
{
    public static BackendAdapterHandler Default => Execute;

    private static readonly HttpClient Http = new();

    private static Dictionary<string, object?> Execute(
        Dictionary<string, object?> capability,
        Dictionary<string, object?> plan,
        Dictionary<string, object?> parameters,
        InvocationContext _context)
    {
        var unresolved = StringList(plan.GetValueOrDefault("unresolved_required_backend_inputs"));
        if (unresolved.Count > 0)
        {
            return Result(capability, plan, "backend_input_incomplete", new() { ["unresolved_required_backend_inputs"] = unresolved });
        }

        var token = Environment.GetEnvironmentVariable("SLACK_BOT_TOKEN")?.Trim() ?? "";
        return Text(capability.GetValueOrDefault("capability_id")) switch
        {
            "slack.channel.read_context" => string.IsNullOrWhiteSpace(token)
                ? Result(capability, plan, "backend_not_configured", new() { ["missing_env"] = "SLACK_BOT_TOKEN" })
                : ReadChannel(capability, plan, parameters, token),
            "slack.thread.summarize" => string.IsNullOrWhiteSpace(token)
                ? Result(capability, plan, "backend_not_configured", new() { ["missing_env"] = "SLACK_BOT_TOKEN" })
                : ReadThread(capability, plan, parameters, token),
            "slack.message.prepare" or "slack.incident_update.prepare" or "slack.announcement.request" => PrepareOrSendMessage(capability, plan, parameters, token, _context),
            _ => Result(capability, plan, "backend_execution_stub", new() { ["note"] = "No Slack custom handler is registered for this capability." }),
        };
    }

    private static Dictionary<string, object?> ReadChannel(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token)
    {
        var channelId = Text(parameters.GetValueOrDefault("channel_id"));
        if (!ChannelAllowed(channelId)) return Restricted(capability, plan, channelId);
        var limit = BoundedLimit(parameters.GetValueOrDefault("limit"), 20, 50);
        var query = Text(parameters.GetValueOrDefault("query")).ToLowerInvariant();
        var payload = Slack(token, "conversations.history", new() { ["channel"] = channelId, ["limit"] = limit });
        if (!Ok(payload)) return BackendError(capability, plan, payload);
        var messages = MessageSummaries(payload.GetValueOrDefault("messages"));
        if (!string.IsNullOrWhiteSpace(query)) messages = messages.Where(message => Text(message.GetValueOrDefault("text")).ToLowerInvariant().Contains(query)).ToList();
        if (messages.Count > limit) messages = messages.Take(limit).ToList();
        return Result(capability, plan, "completed", new() { ["result"] = new Dictionary<string, object?> { ["messages"] = messages, ["count"] = messages.Count, ["channel_id"] = channelId } });
    }

    private static Dictionary<string, object?> ReadThread(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token)
    {
        var channelId = Text(parameters.GetValueOrDefault("channel_id"));
        if (!ChannelAllowed(channelId)) return Restricted(capability, plan, channelId);
        var threadTs = Text(parameters.GetValueOrDefault("thread_ts"));
        var limit = BoundedLimit(parameters.GetValueOrDefault("limit"), 50, 100);
        var payload = Slack(token, "conversations.replies", new() { ["channel"] = channelId, ["ts"] = threadTs, ["limit"] = limit });
        if (!Ok(payload)) return BackendError(capability, plan, payload);
        var messages = MessageSummaries(payload.GetValueOrDefault("messages"));
        return Result(capability, plan, "completed", new() { ["result"] = new Dictionary<string, object?> { ["messages"] = messages, ["count"] = messages.Count, ["channel_id"] = channelId, ["thread_ts"] = threadTs } });
    }

    private static Dictionary<string, object?> PrepareOrSendMessage(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> parameters, string token, InvocationContext context)
    {
        var channelId = Text(parameters.GetValueOrDefault("channel_id"));
        if (!ChannelAllowed(channelId)) return Restricted(capability, plan, channelId);
        var body = new Dictionary<string, object?> { ["channel"] = channelId, ["text"] = MessageText(capability, parameters) };
        if (!string.IsNullOrWhiteSpace(Text(parameters.GetValueOrDefault("thread_ts")))) body["thread_ts"] = Text(parameters.GetValueOrDefault("thread_ts"));
        var preview = Result(capability, plan, "prepared", new()
        {
            ["approval_required"] = true,
            ["mutation_performed"] = false,
            ["slack_action"] = "chat.postMessage",
            ["post_message_request"] = new Dictionary<string, object?> { ["method"] = "POST", ["path"] = "/api/chat.postMessage", ["body"] = body },
            ["note"] = "Prepared a Slack message payload. No Slack message was sent.",
        });
        if (Environment.GetEnvironmentVariable("ANIP_SLACK_ALLOW_SEND") != "true" || string.IsNullOrWhiteSpace(context?.ApprovalGrant)) return preview;
        if (string.IsNullOrWhiteSpace(token))
        {
            preview["execution_status"] = "backend_error";
            preview["slack_error"] = new Dictionary<string, object?> { ["ok"] = false, ["error"] = "missing_slack_token" };
            return preview;
        }
        var posted = Slack(token, "chat.postMessage", body);
        if (!Ok(posted))
        {
            preview["execution_status"] = "backend_error";
            preview["slack_error"] = posted;
            return preview;
        }
        preview["execution_status"] = "completed";
        preview["approval_required"] = false;
        preview["mutation_performed"] = true;
        preview["posted_message"] = new Dictionary<string, object?> { ["channel"] = posted.GetValueOrDefault("channel"), ["ts"] = posted.GetValueOrDefault("ts") };
        preview["approval_grant_id"] = context.ApprovalGrant;
        preview["note"] = "Sent Slack message after the ANIP runtime validated and reserved an approval grant.";
        return preview;
    }

    public static Dictionary<string, object?> Slack(string token, string path, Dictionary<string, object?> body)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, $"https://slack.com/api/{path}");
        request.Headers.Accept.ParseAdd("application/json");
        request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);
        request.Content = new FormUrlEncodedContent(body.Where(item => item.Value is not null).ToDictionary(item => item.Key, item => Text(item.Value)));
        var response = Http.Send(request);
        var text = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
        var payload = string.IsNullOrWhiteSpace(text) ? new Dictionary<string, object?>() : JsonSerializer.Deserialize<Dictionary<string, object?>>(text)!;
        if (!response.IsSuccessStatusCode) return new Dictionary<string, object?> { ["ok"] = false, ["error"] = "slack_http_error", ["status"] = (int)response.StatusCode, ["detail"] = payload };
        return payload;
    }

    private static string MessageText(Dictionary<string, object?> capability, Dictionary<string, object?> parameters)
    {
        return Text(capability.GetValueOrDefault("capability_id")) switch
        {
            "slack.incident_update.prepare" => string.Join("\n", new[]
            {
                $"Incident {Text(parameters.GetValueOrDefault("incident_id"))}: {Text(parameters.GetValueOrDefault("status"))}",
                Text(parameters.GetValueOrDefault("summary")),
                string.IsNullOrWhiteSpace(Text(parameters.GetValueOrDefault("next_update_time"))) ? "" : $"Next update: {Text(parameters.GetValueOrDefault("next_update_time"))}",
            }.Where(item => !string.IsNullOrWhiteSpace(item))),
            "slack.announcement.request" => (string.IsNullOrWhiteSpace(Text(parameters.GetValueOrDefault("audience"))) ? "" : $"[{Text(parameters.GetValueOrDefault("audience"))}] ") + Text(parameters.GetValueOrDefault("announcement")),
            _ => Text(parameters.GetValueOrDefault("text")),
        };
    }

    private static Dictionary<string, object?> Result(Dictionary<string, object?> capability, Dictionary<string, object?> plan, string status, Dictionary<string, object?> extra)
    {
        var result = new Dictionary<string, object?> { ["execution_status"] = status, ["capability_id"] = capability.GetValueOrDefault("capability_id"), ["selected_backend"] = plan.GetValueOrDefault("selected_binding"), ["semantic_input"] = plan.GetValueOrDefault("semantic_input"), ["backend_input_contract"] = plan.GetValueOrDefault("backend_input_contract") };
        foreach (var item in extra) result[item.Key] = item.Value;
        return result;
    }

    private static Dictionary<string, object?> Restricted(Dictionary<string, object?> capability, Dictionary<string, object?> plan, string channelId) =>
        Result(capability, plan, "restricted", new() { ["channel_id"] = channelId, ["reason"] = "Slack channel is outside the configured ANIP channel policy." });

    private static Dictionary<string, object?> BackendError(Dictionary<string, object?> capability, Dictionary<string, object?> plan, Dictionary<string, object?> payload) =>
        Result(capability, plan, "backend_error", new() { ["slack_error"] = payload });

    private static bool ChannelAllowed(string channelId)
    {
        var blocked = CsvEnv("ANIP_SLACK_BLOCKED_CHANNELS");
        var allowed = CsvEnv("ANIP_SLACK_ALLOWED_CHANNELS");
        return !blocked.Contains(channelId) && (allowed.Count == 0 || allowed.Contains(channelId));
    }

    private static List<Dictionary<string, object?>> MessageSummaries(object? value)
    {
        var result = new List<Dictionary<string, object?>>();
        if (value is JsonElement element && element.ValueKind == JsonValueKind.Array)
        {
            foreach (var message in element.EnumerateArray())
            {
                result.Add(new()
                {
                    ["ts"] = Get(message, "ts"),
                    ["user"] = FirstNonEmpty(Get(message, "user"), Get(message, "bot_id")),
                    ["text"] = Get(message, "text"),
                    ["thread_ts"] = Get(message, "thread_ts"),
                });
            }
        }
        return result;
    }

    private static bool Ok(Dictionary<string, object?> payload) =>
        payload.GetValueOrDefault("ok") is bool ok ? ok : payload.GetValueOrDefault("ok") is JsonElement element && element.ValueKind == JsonValueKind.True;

    private static string Get(JsonElement element, string key) => element.TryGetProperty(key, out var value) ? (value.ValueKind == JsonValueKind.String ? value.GetString() ?? "" : value.ToString()) : "";
    private static string FirstNonEmpty(params string[] values) => values.FirstOrDefault(value => !string.IsNullOrWhiteSpace(value)) ?? "";
    private static int BoundedLimit(object? value, int defaultValue, int maximum) => int.TryParse(Text(value), out var parsed) ? Math.Max(1, Math.Min(parsed, maximum)) : defaultValue;
    private static string Text(object? value) => value is null ? "" : value.ToString()!.Trim();
    private static List<string> CsvEnv(string name) => (Environment.GetEnvironmentVariable(name) ?? "").Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries).Distinct().ToList();
    private static List<string> StringList(object? value) => value is IEnumerable<object?> items ? items.Select(Text).Where(item => item.Length > 0).ToList() : [];
}

File: examples/showcase/slack_fronting/custom-code-bundles/slack_fronting_java/src/main/java/{{ANIP_JAVA_PACKAGE_PATH}}/BackendAdapter.java
package {{ANIP_JAVA_PACKAGE_NAME}};

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.anip.service.InvocationContext;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.StringJoiner;

@FunctionalInterface
public interface BackendAdapter {

    Map<String, Object> execute(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> adapterInput, InvocationContext context);

    static BackendAdapter defaultAdapter() {
        return new SlackBackendAdapter();
    }
}

final class SlackBackendAdapter implements BackendAdapter {
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final HttpClient HTTP = HttpClient.newHttpClient();

    @Override
    public Map<String, Object> execute(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, InvocationContext context) {
        List<String> unresolved = stringList(plan.get("unresolved_required_backend_inputs"));
        if (!unresolved.isEmpty()) {
            return result(capability, plan, "backend_input_incomplete", record("unresolved_required_backend_inputs", unresolved));
        }
        String token = System.getenv().getOrDefault("SLACK_BOT_TOKEN", "").trim();
        return switch (text(capability.get("capability_id"))) {
            case "slack.channel.read_context" -> token.isBlank() ? result(capability, plan, "backend_not_configured", record("missing_env", "SLACK_BOT_TOKEN")) : readChannel(capability, plan, params, token);
            case "slack.thread.summarize" -> token.isBlank() ? result(capability, plan, "backend_not_configured", record("missing_env", "SLACK_BOT_TOKEN")) : readThread(capability, plan, params, token);
            case "slack.message.prepare", "slack.incident_update.prepare", "slack.announcement.request" -> prepareOrSendMessage(capability, plan, params, token, context);
            default -> result(capability, plan, "backend_execution_stub", record("note", "No Slack custom handler is registered for this capability."));
        };
    }

    private static Map<String, Object> readChannel(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token) {
        String channelId = text(params.get("channel_id"));
        if (!channelAllowed(channelId)) return restricted(capability, plan, channelId);
        int limit = boundedLimit(params.get("limit"), 20, 50);
        String query = text(params.get("query")).toLowerCase();
        Map<String, Object> payload = slack(token, "conversations.history", record("channel", channelId, "limit", limit));
        if (!Boolean.TRUE.equals(payload.get("ok"))) return backendError(capability, plan, payload);
        List<Map<String, Object>> messages = messageSummaries(payload.get("messages"));
        if (!query.isBlank()) messages.removeIf(message -> !text(message.get("text")).toLowerCase().contains(query));
        if (messages.size() > limit) messages = new ArrayList<>(messages.subList(0, limit));
        return result(capability, plan, "completed", record("result", record("messages", messages, "count", messages.size(), "channel_id", channelId)));
    }

    private static Map<String, Object> readThread(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token) {
        String channelId = text(params.get("channel_id"));
        if (!channelAllowed(channelId)) return restricted(capability, plan, channelId);
        String threadTs = text(params.get("thread_ts"));
        int limit = boundedLimit(params.get("limit"), 50, 100);
        Map<String, Object> payload = slack(token, "conversations.replies", record("channel", channelId, "ts", threadTs, "limit", limit));
        if (!Boolean.TRUE.equals(payload.get("ok"))) return backendError(capability, plan, payload);
        List<Map<String, Object>> messages = messageSummaries(payload.get("messages"));
        return result(capability, plan, "completed", record("result", record("messages", messages, "count", messages.size(), "channel_id", channelId, "thread_ts", threadTs)));
    }

    private static Map<String, Object> prepareOrSendMessage(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String token, InvocationContext context) {
        String channelId = text(params.get("channel_id"));
        if (!channelAllowed(channelId)) return restricted(capability, plan, channelId);
        Map<String, Object> body = record("channel", channelId, "text", messageText(capability, params));
        if (!text(params.get("thread_ts")).isBlank()) body.put("thread_ts", text(params.get("thread_ts")));
        Map<String, Object> preview = result(capability, plan, "prepared", record(
            "approval_required", true,
            "mutation_performed", false,
            "slack_action", "chat.postMessage",
            "post_message_request", record("method", "POST", "path", "/api/chat.postMessage", "body", body),
            "note", "Prepared a Slack message payload. No Slack message was sent."
        ));
        String approvalGrant = context == null ? "" : text(context.getApprovalGrant());
        if (!"true".equals(System.getenv().getOrDefault("ANIP_SLACK_ALLOW_SEND", "")) || approvalGrant.isBlank()) return preview;
        if (token.isBlank()) {
            preview.put("execution_status", "backend_error");
            preview.put("slack_error", record("ok", false, "error", "missing_slack_token"));
            return preview;
        }
        Map<String, Object> posted = slack(token, "chat.postMessage", body);
        if (!Boolean.TRUE.equals(posted.get("ok"))) {
            preview.put("execution_status", "backend_error");
            preview.put("slack_error", posted);
            return preview;
        }
        preview.put("execution_status", "completed");
        preview.put("approval_required", false);
        preview.put("mutation_performed", true);
        preview.put("posted_message", record("channel", posted.get("channel"), "ts", posted.get("ts")));
        preview.put("approval_grant_id", approvalGrant);
        preview.put("note", "Sent Slack message after the ANIP runtime validated and reserved an approval grant.");
        return preview;
    }

    static Map<String, Object> slack(String token, String path, Map<String, Object> body) {
        try {
            StringJoiner form = new StringJoiner("&");
            for (Map.Entry<String, Object> entry : body.entrySet()) form.add(url(entry.getKey()) + "=" + url(text(entry.getValue())));
            HttpRequest request = HttpRequest.newBuilder(URI.create("https://slack.com/api/" + path))
                .header("Accept", "application/json")
                .header("Authorization", "Bearer " + token)
                .header("Content-Type", "application/x-www-form-urlencoded")
                .POST(HttpRequest.BodyPublishers.ofString(form.toString()))
                .build();
            HttpResponse<String> response = HTTP.send(request, HttpResponse.BodyHandlers.ofString());
            Map<String, Object> payload = response.body().isBlank() ? new LinkedHashMap<>() : MAPPER.readValue(response.body(), new TypeReference<>() {});
            if (response.statusCode() < 200 || response.statusCode() >= 300) return record("ok", false, "error", "slack_http_error", "status", response.statusCode(), "detail", payload);
            return payload;
        } catch (Exception exc) {
            return record("ok", false, "error", "slack_http_error", "detail", exc.toString());
        }
    }

    private static String messageText(Map<String, Object> capability, Map<String, Object> params) {
        return switch (text(capability.get("capability_id"))) {
            case "slack.incident_update.prepare" -> String.join("\n", nonEmpty(List.of(
                "Incident " + text(params.get("incident_id")) + ": " + text(params.get("status")),
                text(params.get("summary")),
                text(params.get("next_update_time")).isBlank() ? "" : "Next update: " + text(params.get("next_update_time"))
            )));
            case "slack.announcement.request" -> (text(params.get("audience")).isBlank() ? "" : "[" + text(params.get("audience")) + "] ") + text(params.get("announcement"));
            default -> text(params.get("text"));
        };
    }

    @SuppressWarnings("unchecked")
    private static List<Map<String, Object>> messageSummaries(Object value) {
        List<Map<String, Object>> result = new ArrayList<>();
        if (value instanceof List<?> items) {
            for (Object item : items) {
                if (item instanceof Map<?, ?> message) {
                    Map<String, Object> typed = (Map<String, Object>) message;
                    result.add(record("ts", typed.get("ts"), "user", firstNonBlank(typed.get("user"), typed.get("bot_id")), "text", typed.get("text"), "thread_ts", typed.get("thread_ts")));
                }
            }
        }
        return result;
    }

    private static Map<String, Object> result(Map<String, Object> capability, Map<String, Object> plan, String status, Map<String, Object> extra) {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("execution_status", status);
        result.put("capability_id", capability.get("capability_id"));
        result.put("selected_backend", plan.get("selected_binding"));
        result.put("semantic_input", plan.get("semantic_input"));
        result.put("backend_input_contract", plan.get("backend_input_contract"));
        result.putAll(extra);
        return result;
    }

    private static Map<String, Object> restricted(Map<String, Object> capability, Map<String, Object> plan, String channelId) {
        return result(capability, plan, "restricted", record("channel_id", channelId, "reason", "Slack channel is outside the configured ANIP channel policy."));
    }

    private static Map<String, Object> backendError(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> payload) {
        return result(capability, plan, "backend_error", record("slack_error", payload));
    }

    private static boolean channelAllowed(String channelId) {
        List<String> blocked = csvEnv("ANIP_SLACK_BLOCKED_CHANNELS");
        List<String> allowed = csvEnv("ANIP_SLACK_ALLOWED_CHANNELS");
        return !blocked.contains(channelId) && (allowed.isEmpty() || allowed.contains(channelId));
    }

    private static List<String> csvEnv(String name) {
        List<String> result = new ArrayList<>();
        for (String item : System.getenv().getOrDefault(name, "").split(",")) {
            String value = item.trim();
            if (!value.isBlank() && !result.contains(value)) result.add(value);
        }
        return result;
    }

    private static int boundedLimit(Object value, int defaultValue, int max) {
        try {
            return Math.max(1, Math.min(Integer.parseInt(text(value)), max));
        } catch (Exception ignored) {
            return defaultValue;
        }
    }

    private static List<String> stringList(Object value) {
        List<String> result = new ArrayList<>();
        if (value instanceof List<?> items) for (Object item : items) result.add(text(item));
        return result;
    }

    private static List<String> nonEmpty(List<String> values) {
        List<String> result = new ArrayList<>();
        for (String value : values) if (!value.isBlank()) result.add(value);
        return result;
    }

    private static String firstNonBlank(Object... values) {
        for (Object value : values) if (!text(value).isBlank()) return text(value);
        return "";
    }

    private static String text(Object value) {
        return value == null ? "" : value.toString().trim();
    }

    private static String url(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    private static Map<String, Object> record(Object... pairs) {
        Map<String, Object> result = new LinkedHashMap<>();
        for (int index = 0; index + 1 < pairs.length; index += 2) result.put(String.valueOf(pairs[index]), pairs[index + 1]);
        return result;
    }
}
